### PR TITLE
Allow to build LexBuffer from array without copying.

### PR DESCRIPTION
**Pull request by smoothdeveloper**
All usages of FromChars seem to be taking a throw away array that caller don't keep a reference, usages found:

src\fsharp\lex.fsl(155)
src\fsharp\UnicodeLexing.fs(18)
src\fsharp\UnicodeLexing.fs(68)
src\fsharp\FSharp.Compiler.Private\lex.fs(154)
tests\FSharp.Compiler.UnitTests\HashIfExpression.fs(71)

### CODE REVIEW

File: src/utils/prim-lexing.fs [hunk 1]
```diff
@@ -156,14 +156,21 @@ namespace Internal.Utilities.Text.Lexing
             new LexBuffer<'Char>(filler)
               
         // A full type signature is required on this method because it is used at more specific types within its own scope
-        static member FromArray (s: 'Char[]) : LexBuffer<'Char> = 
+        // Important: This method takes ownership of the array
```
Comment (forki) [inline src/utils/prim-lexing.fs]: ask the borrow checker ;-)

File: src/utils/prim-lexing.fs [hunk 2]
```diff
@@ -156,14 +156,21 @@ namespace Internal.Utilities.Text.Lexing
...
+        static member FromArrayNoCopy (buffer: 'Char[]) : LexBuffer<'Char> = 
             let lexBuffer = new LexBuffer<'Char>(fun _ -> ())
-            let buffer = Array.copy s
```
Comment (forki) [inline src/utils/prim-lexing.fs]: please remove the ; while you are touching this

File: src/utils/prim-lexing.fsi [hunk 3]
```diff
@@ -65,6 +65,7 @@ type internal LexBuffer<'Char> =
     member IsPastEndOfStream: bool with get,set
```
Comment (forki) [inline src/utils/prim-lexing.fsi]: FromArryNoCopy should go here

Comment (smoothdeveloper) [reply src/utils/prim-lexing.fsi]: Notice that `FromArray` isn't exposed in this file, this is why I haven't added the extra method.

Comment (forki) [reply src/utils/prim-lexing.fsi]: ok thanks

File: src/utils/prim-lexing.fs [hunk 4]
```diff
@@ -156,14 +156,21 @@ namespace Internal.Utilities.Text.Lexing
...
             let lexBuffer = new LexBuffer<'Char>(fun _ -> ())
-            let buffer = Array.copy s 
-            lexBuffer.Buffer <- buffer;
-            lexBuffer.BufferMaxScanLength <- buffer.Length;
+            lexBuffer.Buffer <- buffer
+            lexBuffer.BufferMaxScanLength <- buffer.Length
             lexBuffer
 
-        static member FromChars (arr:char[]) = LexBuffer.FromArray arr 
+        // A full type signature is required on this method because it is used at more specific types within its own scope
```
Comment (forki) [inline src/utils/prim-lexing.fs]: I think the comment about full signature can go.

File: src/utils/prim-lexing.fs (hunk 2 above)
Comment (smoothdeveloper) [reply src/utils/prim-lexing.fs]: Done, were left due to oversight.

File: src/utils/prim-lexing.fs (hunk 4 above)
Comment (smoothdeveloper) [reply src/utils/prim-lexing.fs]: Waiting for further confirmation, if I should remove all the same comments on the two other methods as well.

Notice that there are dubious comments here:

https://github.com/Microsoft/visualfsharp/blob/75e435e74b81bb47c4a0ced11be0d7560d2a66ea/src/fsharp/UnicodeLexing.fs#L35-L40

It doesn't look like those comments apply as the stream is totally handled within the method and we just return the LexBuffer object.

**Comment by smoothdeveloper** [comment]
I have no clue if this failing test is broken because of the change:

```
=> FSharp.Core.UnitTests.FSharp_Core.Type_Forwarding.TypeForwardingModule.TypeForwarding
currentRuntimeVersion = v4.0.30319; currentFSharpCoreTargetRuntime=v4.0.30319 tupleAssemblyName=mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
-----------------------------------------------------------------
-----------------------------------------------------------------
Error: Running tests net40-coreunit failed, see logs above -- FAILED
-----------------------------------------------------------------
```

If it is, I'd somewhat expect more tests to be broken, not just one.

**Comment by dsyme** [comment]
The failing test is this one:
```
1) Failed : FSharp.Core.UnitTests.FSharp_Core.Microsoft_FSharp_Control.AsyncModule.SleepContinuations
  Expected: 0
  But was:  1
```

**Comment by dsyme** [comment]
@smoothdeveloper I think that may have been a transient failure due to timing.  I was worried the recent control.fs cleanup had caused it but the test is passing on my machine

**Comment by dsyme** [comment]
@dotnet-bot test Windows_NT Release_ci_part2 Build please

File: src/utils/prim-lexing.fs (hunk 4 above)
Comment (dsyme) [reply src/utils/prim-lexing.fs]: Yeah those comments look old

**Pull request merged**